File: tests/generic.py
# pip3 install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cu124

import numpy as np;

import torch
import torch.nn as nn
import torch.nn.functional as F

from common import pretty_logits

def eval_net(net, W, input):
    net.eval()
    logits = net(input)
    print(f"logits:")
    print(pretty_logits(logits, 0xFFFFFFFF))
    print("weight:")
    print(pretty_logits(W.weight, 0xFFFFFFFF))
    if W.weight.grad is not None:
        print("weight.grad:")
        print(pretty_logits(W.weight.grad, 0xFFFFFFFF))
    if W.bias is not None:
        print("bias:")
        print(pretty_logits(W.bias, 0xFFFFFFFF))
        if W.bias.grad is not None:
            print("bias.grad:")
            print(pretty_logits(W.bias.grad, 0xFFFFFFFF))

def test_net(optim, loss, lr, bias, activation, momentum, weight_decay, decimals):
    I = 5
    H = 7
    O = 3
    B = 4

    print("<test_net>")

    torch.manual_seed(137)
    W_hidden = nn.Linear(I, H, bias=bias)
    W_output = nn.Linear(H, O, bias=bias)

    if activation == "Identity":
        F_act = nn.Identity()
    elif activation == "ReLU":
        F_act = nn.ReLU()
    elif activation == "Sigmoid":
        F_act = nn.Sigmoid()
    elif activation == "LeakyReLU":
        F_act = nn.LeakyReLU()
    elif activation == "Dropout":
        F_act = nn.Dropout()
    elif activation == "Tanh":
        F_act = nn.Tanh()

    net = nn.Sequential(
        W_hidden,
        F_act,
        W_output,
        nn.Sigmoid() if loss == "BCELoss" else nn.Identity());

    sample = torch.ones(B, I)
    target = torch.ones(B, O)

    eval_net(net, W_output, sample)

    if optim == "SGD":
        optimizer = torch.optim.SGD(net.parameters(), lr = lr, momentum=momentum, weight_decay=weight_decay)
    elif optim == "AdamW":
        optimizer = torch.optim.AdamW(net.parameters(), lr = lr, weight_decay=weight_decay)

    for step in range(1000):
        net.train()
        logits = net(sample)
        if loss == "MSELoss":
            diff = F.mse_loss(logits, target)
        elif loss == "BCELoss":
            diff = F.binary_cross_entropy(logits, target)
        optimizer.zero_grad()
        diff.backward()
        optimizer.step()
        if step % 100 == 0:
            print(f"[{step}]: {pretty_logits(diff, decimals=decimals)}")

    eval_net(net, W_output, sample)
    print("</test_net>")

if __name__ == "__main__":
    import argparse

    from pathlib import Path

    path = Path(__file__).parent.absolute()

    parser = argparse.ArgumentParser()
    parser.add_argument("--bias", type=str, default="yes")
    parser.add_argument("--lr", type=float, default=1e-3)
    parser.add_argument("--m", type=float, default=1e-1)
    parser.add_argument("--weight_decay", type=float, default=1e-1)
    parser.add_argument("--a", type=str, default="Identity")
    parser.add_argument("--o", type=str, default="SGD")
    parser.add_argument("--d", type=int, default=4)
    parser.add_argument("--loss", type=str, default="MSELoss")
    
    args = parser.parse_args()

    assert args.a in { "Identity", "ReLU", "Sigmoid", "Tanh", "LeakyReLU", "Dropout" }
    assert args.o in { "SGD", "AdamW" }
    assert args.loss in { "MSELoss", "BCELoss" }
    assert args.bias in { "yes", "no" }
    assert 0 <= args.lr <= 1
    assert 0 <= args.m <= 1
    assert 0 <= args.weight_decay <= 1

    test_net(optim = args.o,
             loss=args.loss,
             lr = args.lr,
             bias = True if args.bias == "yes" else False,
             activation = args.a,
             momentum = args.m,
             weight_decay = args.weight_decay,
             decimals=args.d);

File: tests/generic.cs
using System;
using System.IO;

using nn;

unsafe internal static class generic {
    // static void reset_weights(Linear layer, IRNG g, string nonlinearity = "leaky_relu") {
    //     nn.init.kaiming_uniform_(
    //         layer._Weight.data,
    //         layer._Weight.numel(),
    //         g,
    //         layer.I,
    //         (float)Math.Sqrt(5),
    //         nonlinearity);
    // 
    //     if (layer._Bias != null) {
    //         nn.rand.uniform_(
    //             layer._Bias.data,
    //             layer._Bias.numel(),
    //             g,
    //             -(float)(1.0 / Math.Sqrt(layer.I)),
    //             (float)(1.0 / Math.Sqrt(layer.I)));
    //     }
    // }

    public static void eval_net(TextWriter Console, IModel net, Linear W, Tensor input, bool grad) {
        net.eval();
        var logits = net.forward(input);
        Console.WriteLine("logits:");
        Console.WriteLine(Common.pretty_logits(logits.data, logits.numel(), 0xFFFFFFFF));
        Console.WriteLine("weight:");
        Console.WriteLine(Common.pretty_logits(W._Weight.data, W._Weight.numel(), 0xFFFFFFFF));
        if (W._Weight.grad != null && grad) {
            Console.WriteLine("weight.grad:");
            Console.WriteLine(Common.pretty_logits(W._Weight.grad, W._Weight.numel(), 0xFFFFFFFF));
        }
        if (W._Bias != null) {
            Console.WriteLine("bias:");
            Console.WriteLine(Common.pretty_logits(W._Bias.data, W._Bias.numel(), 0xFFFFFFFF));
            if (W._Bias.grad != null && grad) {
                Console.WriteLine("bias.grad:");
                Console.WriteLine(Common.pretty_logits(W._Bias.grad, W._Bias.numel(), 0xFFFFFFFF));
            }
        }
    }

    public static void test_net(TextWriter Console,
        bool bias,
        string activation,
        string optim,
        string loss,
        float lr,
        float momentum,
        float weight_decay,
        int decimals) {

        const int I = 5;
        const int H = 7;
        const int O = 3;
        const int B = 4;

        Console.WriteLine($"<test_net>");

        var g = new nn.rand.mt19937(137);
        Linear W_hidden = new nn.Linear(I, H, bias: bias, Linear.Kernel.Naive);
        Linear W_output = new nn.Linear(H, O, bias: bias, Linear.Kernel.Naive);
        nn.init.reset_weights(W_hidden, "leaky_relu", g);
        nn.init.reset_weights(W_output, "leaky_relu", g);
        IModel F_act = new nn.Identity();
        switch (activation) {
            case "Identity":
                break;
            case "Tanh":
                F_act = new nn.Tanh();
                break;
            case "LeakyReLU":
                F_act = new nn.LeakyReLU();
                break;
            case "Dropout":
                F_act = new nn.Dropout(g);
                break;
            case "ReLU":
                F_act = new nn.ReLU();
                break;
            case "Sigmoid":
                F_act = new nn.Sigmoid();
                break;
            default:
                if (!string.IsNullOrEmpty(activation)) {
                    throw new ArgumentOutOfRangeException($"The specified activation '{activation}' is not supported");
                }
                break;
        }

        var net = new nn.Sequential(
            W_hidden,
            F_act,
            W_output,
            loss == "BCELoss" ? new Sigmoid() : new Identity());

        var input = Tensor.ones((uint)I * B, requires_grad: true);
        var target = Tensor.ones((uint)O * B, requires_grad: false);
        eval_net(Console, net, W_output, input, grad: false);

        IOptimizer optimizer = null;
        if (optim == "SGD") {
            optimizer = new nn.SGD(net.parameters(), lr: lr, momentum: momentum, weight_decay: weight_decay);
        } else if (optim == "AdamW") {
            optimizer = new nn.AdamW(net.parameters(), lr: lr, weight_decay: weight_decay);
        } else {
            throw new ArgumentOutOfRangeException($"The specified optimizer '{optim}' is not supported");
        }
        for (int step = 0; step < 1000; step++) {
            net.train();
            var logits = net.forward(input);
            var diff = 0.0;
            if (loss == "MSELoss") {
                diff = F.mse_loss(
                    logits.data,
                    logits.grad,
                    logits.numel(),
                    target.data);
            } else if (loss == "BCELoss") {
                diff = F.binary_cross_entropy(
                    logits,
                    target);
            } else {
                throw new ArgumentOutOfRangeException($"The specified loss '{loss}' is not supported");
            }
            optimizer.zero_grad();
            net.backward(logits);
            optimizer.step();
            if (step % 100 == 0) {
                Console.WriteLine($"[{step}]: [{Math.Round(diff, decimals):f4}]");
            }
        }

        eval_net(Console, net, W_output, input, grad: true);
        Console.WriteLine("</test_net>");
    }
}